compiler/packages/babel-plugin-react-compiler/src/HIR/PropagateScopeDependenciesHIR.ts

```diff
@@ -0,0 +1,512 @@
+      tree.addDependency({...dep});
+    }
+    /**
+     * Step 2: Mark hoistable dependencies, given the basic block in
+     * which the scope begins.
+     */
+    recordHoistablePropertyReads(nodes, scope.id, tree);
+    const candidates = tree.deriveMinimalDependencies();
+    for (const candidateDep of candidates) {
+      if (
+        !Iterable_some(
+          scope.dependencies,
+          existingDep =>
+            existingDep.identifier.declarationId ===
+              candidateDep.identifier.declarationId &&
+            areEqualPaths(existingDep.path, candidateDep.path),
+        )
+      )
+        scope.dependencies.add(candidateDep);
+    }
+  }
+}
+
+function findTemporariesUsedOutsideDeclaringScope(
+  fn: HIRFunction,
+): Set<DeclarationId> {
+  /*
+   * tracks all relevant LoadLocal and PropertyLoad lvalues
+   * and the scope where they are defined
+   */
+  const declarations = new Map<DeclarationId, ScopeId>();
+  const prunedScopes = new Set<ScopeId>();
+  const scopeTraversal = new ScopeBlockTraversal();
+  const usedOutsideDeclaringScope = new Set<DeclarationId>();
+
+  function handlePlace(place: Place): void {
+    const declaringScope = declarations.get(place.identifier.declarationId);
+    if (
+      declaringScope != null &&
+      !scopeTraversal.isScopeActive(declaringScope) &&
+      !prunedScopes.has(declaringScope)
+    ) {
+      // Declaring scope is not active === used outside declaring scope
+      usedOutsideDeclaringScope.add(place.identifier.declarationId);
+    }
+  }
+
+  function handleInstruction(instr: Instruction): void {
+    const scope = scopeTraversal.currentScope;
+    if (scope == null || prunedScopes.has(scope)) {
+      return;
+    }
+    switch (instr.value.kind) {
+      case 'LoadLocal':
+      case 'LoadContext':
+      case 'PropertyLoad': {
+        declarations.set(instr.lvalue.identifier.declarationId, scope);
+        break;
+      }
+      default: {
+        break;
+      }
+    }
+  }
+
+  for (const [blockId, block] of fn.body.blocks) {
+    scopeTraversal.recordScopes(block);
+    const scopeStartInfo = scopeTraversal.blockInfos.get(blockId);
+    if (scopeStartInfo?.kind === 'begin' && scopeStartInfo.pruned) {
+      prunedScopes.add(scopeStartInfo.scope.id);
+    }
+    for (const instr of block.instructions) {
+      for (const place of eachInstructionOperand(instr)) {
+        handlePlace(place);
+      }
+      handleInstruction(instr);
+    }
+
+    for (const place of eachTerminalOperand(block.terminal)) {
+      handlePlace(place);
+    }
+  }
+  return usedOutsideDeclaringScope;
+}
+
+type Decl = {
+  id: InstructionId;
+  scope: Stack<ReactiveScope>;
+};
+
+class Context {
+  #temporariesUsedOutsideScope: Set<DeclarationId>;
+  #declarations: Map<DeclarationId, Decl> = new Map();
+  #reassignments: Map<Identifier, Decl> = new Map();
+  // Reactive dependencies used in the current reactive scope.
+  #dependencies: Stack<Array<ReactiveScopeDependency>> = empty();
```

Comment:
Note that `dependencies is now a stack since we no longer are recursing into scopes. We only ever use the top entry of this, so I'm happy to add/use a new array based utility class that only exposes push, pop, and top